### PR TITLE
pg_dump: set correct numsegments for relations

### PR DESCRIPTION
**Pull request by zhangjackey**
pg_dump dump relations as CREATE commands, however the internal
numsegments property cannot be specified in CREATE commands, their
values are decided on execution from the cluster size, however as a
utility mode master is used to restore the tables the invalid cluster
size is used as numsegments for all the tables.

To fix the issue we generate an UPDATE gp_distribution_policy command
after each CREATE command to correct the numsegments property, so no
matter there are partially distributed tables or not they can always be
correctly restored.

A new UDF which is used to check expansion status is provided,
for more details please refer the commit message.

## Background
Previous discussion can be found in https://github.com/greenplum-db/gpdb/pull/6326 and https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/Jfbj3DQgn0w/SKwo3PYaBQAJ, this PR is to fix the issue in binary upgrade.  The check of expansion status will be in a separate PR, but we also have included a temporary one in this PR, once that PR is ready we could replace the temporary one with it.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [x] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`

### CODE REVIEW

File: src/bin/pg_dump/pg_dump.c [hunk 1]
```diff
@@ -14062,6 +14063,13 @@ dumpExternal(Archive *fout, TableInfo *tbinfo, PQExpBuffer q, PQExpBuffer delq)
 
 		appendPQExpBufferStr(q, ";\n");
 
+		if (iswritable && binary_upgrade)
```
[the remark this answers is not in the document]
Comment (zhangjackey) [reply src/bin/pg_dump/pg_dump.c]: OK, Thanks.

File: src/bin/pg_dump/pg_dump.c [hunk 2]
```diff
@@ -17064,6 +17090,76 @@ addDistributedBy(Archive *fout, PQExpBuffer q, TableInfo *tbinfo, int actual_att
...
+ *
+ *  update the numsegments of the passed in relation.  Must be executed after
+ *  the CREATE TABLE command.
+ */
+static void
+updateNumsegments(Archive *fout, PQExpBuffer q, TableInfo *tbinfo)
+{
+	Assert(isGPDB6000OrLater(fout));
+	Assert(binary_upgrade);
+	Assert(tbinfo->numsegments > 0);
```
Comment (danielgustafsson) [inline src/bin/pg_dump/pg_dump.c]: Let's make these actual runtime checks and not assertions.

File: src/bin/pg_dump/pg_dump.c [hunk 3]
```diff
@@ -17064,6 +17090,76 @@ addDistributedBy(Archive *fout, PQExpBuffer q, TableInfo *tbinfo, int actual_att
...
+			{
+				Oid part_oid = atooid(PQgetvalue(partres, i, 0));
+
+				appendPQExpBuffer(q, "%u", part_oid);
+
+				if (count - 1 != i)
+					appendPQExpBufferStr(q, ", ");
+			}
+
+			appendPQExpBuffer(q, ");\n");
```
Comment (danielgustafsson) [inline src/bin/pg_dump/pg_dump.c]: appendPQExpBufferStr()

File: contrib/pg_upgrade/check_gp.c [hunk 4]
```diff
@@ -32,12 +33,93 @@ static void check_orphaned_toastrels(void);
...
+		int			ntups;
+		DbInfo	   *active_db = &old_cluster.dbarr.dbs[dbnum];
+		PGconn	   *conn;
+
+		conn = connectToServer(&old_cluster, active_db->db_name);
+		res = executeQueryOrDie(conn,
+								"SELECT true AS expansion "
+								"FROM gp_distribution_policy d "
+								"JOIN gp_toolkit.__gp_number_of_segments n "
+								"ON d.numsegments <> n.numsegments "
```
[the remark this answers is not in the document]
Comment (danielgustafsson) [reply contrib/pg_upgrade/check_gp.c]: Ehm ..just noticed the comment above now, so I'll answer myself here: yes.

**Comment by hlinnaka** [comment]
If you check that there is no online expansion in progress, and refuse the upgrade if there is, do you need the code in pg_dump to modify `numsegments`? There shouldn't be any tables with a different `numsegments` value when we create the dump, because we already checked for that earlier.

**Comment by pivotal-ning-yu** [comment]
On Fri, Dec 21, 2018 at 6:38 AM Heikki Linnakangas <notifications@github.com>
wrote:

> If you check that there is no online expansion in progress, and refuse the
> upgrade if there is, do you need the code in pg_dump to modify numsegments?
> There shouldn't be any tables with a different numsegments value when we
> create the dump, because we already checked for that earlier.
>

The problem is in binary upgrade mode, in this mode we connect to each
segment in utility mode and restore the schemas, CREATE TABLE sets the
table's numsegments to cluster size, but in cluster size we does not the
correct cluster size.  So far we report cluster size as -1 in utility mode,
so all tables' numsegments will be set to 1, however the data are reused
and were distributed as real cluster size, so there is a mismatch.

One attempt to fix the issue is to let gpdb reports correct cluster size
even in utility mode.  Cluster size is counted from the catalog table
gp_segment_configuration, however it's only valid on master but empty on
segments, so there is no way to get the correct count on utility mode
segments.

That's why we have to manually hack the numsegments property after each
CREATE TABLE.

**Comment by ashwinstar** [comment]
On Sun, Dec 23, 2018 at 4:46 PM Ning Yu <notifications@github.com> wrote:

> On Fri, Dec 21, 2018 at 6:38 AM Heikki Linnakangas <
> notifications@github.com>
> wrote:
>
> > If you check that there is no online expansion in progress, and refuse
> the
> > upgrade if there is, do you need the code in pg_dump to modify
> numsegments?
> > There shouldn't be any tables with a different numsegments value when we
> > create the dump, because we already checked for that earlier.
> >
>
> The problem is in binary upgrade mode, in this mode we connect to each
> segment in utility mode and restore the schemas, CREATE TABLE sets the
> table's numsegments to cluster size, but in cluster size we does not the
> correct cluster size. So far we report cluster size as -1 in utility mode,
> so all tables' numsegments will be set to 1, however the data are reused
> and were distributed as real cluster size, so there is a mismatch.
>
> One attempt to fix the issue is to let gpdb reports correct cluster size
> even in utility mode. Cluster size is counted from the catalog table
> gp_segment_configuration, however it's only valid on master but empty on
> segments, so there is no way to get the correct count on utility mode
> segments.
>

In binary upgrade mode, schema is restored only on master first in utility
mode. Then the master's data directory is copied to segments. Hence, there
shouldn't any need for creating tables on segments in utility mode. So,
having the behavior of using gp_segment_configuration if info is available
via it would be good to cover for upgrade and hack won't be needed.

File: src/bin/pg_dump/pg_dump.c [hunk 5]
```diff
@@ -17033,6 +17044,21 @@ addDistributedBy(Archive *fout, PQExpBuffer q, TableInfo *tbinfo, int actual_att
...
 
+		if (isGPDB6000OrLater(fout))
+		{
+			char	   *numsegments_str;
+
+			Assert(!PQgetisnull(res, 0, 2));
+
+			numsegments_str = PQgetvalue(res, 0, 2);
+
+			Assert(numsegments_str[0]);
```
[the remark this answers is not in the document]
Comment (zhangjackey) [reply src/bin/pg_dump/pg_dump.c]: We can preserve it to check if it is an empty string.

File: src/bin/pg_dump/pg_dump.c (hunk 2 above)
Comment (zhangjackey) [reply src/bin/pg_dump/pg_dump.c]: OK

File: src/bin/pg_dump/pg_dump.c (hunk 3 above)
Comment (zhangjackey) [reply src/bin/pg_dump/pg_dump.c]: OK

**Comment by pivotal-ning-yu** [comment]
On Mon, Dec 24, 2018 at 8:53 AM Ashwin Agrawal <notifications@github.com>
wrote:

> On Sun, Dec 23, 2018 at 4:46 PM Ning Yu <notifications@github.com> wrote:
>
> > On Fri, Dec 21, 2018 at 6:38 AM Heikki Linnakangas <
> > notifications@github.com>
> > wrote:
> >
> > > If you check that there is no online expansion in progress, and refuse
> > the
> > > upgrade if there is, do you need the code in pg_dump to modify
> > numsegments?
> > > There shouldn't be any tables with a different numsegments value when
> we
> > > create the dump, because we already checked for that earlier.
> > >
> >
> > The problem is in binary upgrade mode, in this mode we connect to each
> > segment in utility mode and restore the schemas, CREATE TABLE sets the
> > table's numsegments to cluster size, but in cluster size we does not the
> > correct cluster size. So far we report cluster size as -1 in utility
> mode,
> > so all tables' numsegments will be set to 1, however the data are reused
> > and were distributed as real cluster size, so there is a mismatch.
> >
> > One attempt to fix the issue is to let gpdb reports correct cluster size
> > even in utility mode. Cluster size is counted from the catalog table
> > gp_segment_configuration, however it's only valid on master but empty on
> > segments, so there is no way to get the correct count on utility mode
> > segments.
> >
>
> In binary upgrade mode, schema is restored only on master first in utility
> mode. Then the master's data directory is copied to segments. Hence, there
> shouldn't any need for creating tables on segments in utility mode. So,
> having the behavior of using gp_segment_configuration if info is available
> via it would be good to cover for upgrade and hack won't be needed.
>

Thanks for the explanation.

I made some experiments again, it's gpcheckcat who connects to segments in
utility mode. By adding some condition checks and removing some asserts and
error checks I finally passed the test_gpdb.sh test w/o the pg_upgrade &
pg_dump changes in this PR.

However my concerns on this approach is that gpcheckcat & pg_upgrade &
pg_dump are not the only components who access in utility mode, for other
components whether the hacks are suitable must be carefully checked.
Considering that we are working on __known__ & __existing__ integration
issues between gpexpand and other components, I think it is better not to
introduce new potential integration breakers.

But it's worthy to come back later to this approach.

**Comment by danielgustafsson** [comment]
> If pg_upgrade ensures that all the segments have the default numsegments value, you don't need the pg_dump changes to dump them.

Correct me if I'm wrong, but since the schema restore operation will run in utility mode it will set numsegments to -1, so the dump file would still need the logic to correct it?

**Comment by ashwinstar** [comment]
On Thu, Jan 3, 2019 at 6:02 AM Daniel Gustafsson <notifications@github.com>
wrote:

> If pg_upgrade ensures that all the segments have the default numsegments
> value, you don't need the pg_dump changes to dump them.
>
> Correct me if I'm wrong, but since the schema restore operation will run
> in utility mode it will set numsegments to -1, so the dump file would still
> need the logic to correct it?
>

Not sure its done or not, but had proposed that even in utility mode if
have access to gp_segment_configuration then it should be used for
numsegments and not default to -1. That should fix the pg_upgrade problem
and need no syntax changes in pg_dump for the same.

>

**Comment by pivotal-ning-yu** [comment]
On Fri, Jan 4, 2019 at 12:15 AM Ashwin Agrawal <notifications@github.com>
wrote:

> On Thu, Jan 3, 2019 at 6:02 AM Daniel Gustafsson <notifications@github.com
> >
> wrote:
>
> > If pg_upgrade ensures that all the segments have the default numsegments
> > value, you don't need the pg_dump changes to dump them.
> >
> > Correct me if I'm wrong, but since the schema restore operation will run
> > in utility mode it will set numsegments to -1, so the dump file would
> still
> > need the logic to correct it?
> >
>
> Not sure its done or not, but had proposed that even in utility mode if
> have access to gp_segment_configuration then it should be used for
> numsegments and not default to -1. That should fix the pg_upgrade problem
> and need no syntax changes in pg_dump for the same.
>
Please find my previous reply on this approach:
https://github.com/greenplum-db/gpdb/pull/6518#issuecomment-449826992

In short I think this approach needs more verification, we would come back
to it later.

**Comment by ashwinstar** [comment]
On Tue, Dec 25, 2018 at 12:58 AM Ning Yu <notifications@github.com> wrote:

> On Mon, Dec 24, 2018 at 8:53 AM Ashwin Agrawal <notifications@github.com>
> wrote:
>
> > On Sun, Dec 23, 2018 at 4:46 PM Ning Yu <notifications@github.com>
> wrote:
> >
> > > On Fri, Dec 21, 2018 at 6:38 AM Heikki Linnakangas <
> > > notifications@github.com>
> > > wrote:
> > >
> > > > If you check that there is no online expansion in progress, and
> refuse
> > > the
> > > > upgrade if there is, do you need the code in pg_dump to modify
> > > numsegments?
> > > > There shouldn't be any tables with a different numsegments value when
> > we
> > > > create the dump, because we already checked for that earlier.
> > > >
> > >
> > > The problem is in binary upgrade mode, in this mode we connect to each
> > > segment in utility mode and restore the schemas, CREATE TABLE sets the
> > > table's numsegments to cluster size, but in cluster size we does not
> the
> > > correct cluster size. So far we report cluster size as -1 in utility
> > mode,
> > > so all tables' numsegments will be set to 1, however the data are
> reused
> > > and were distributed as real cluster size, so there is a mismatch.
> > >
> > > One attempt to fix the issue is to let gpdb reports correct cluster
> size
> > > even in utility mode. Cluster size is counted from the catalog table
> > > gp_segment_configuration, however it's only valid on master but empty
> on
> > > segments, so there is no way to get the correct count on utility mode
> > > segments.
> > >
> >
> > In binary upgrade mode, schema is restored only on master first in
> utility
> > mode. Then the master's data directory is copied to segments. Hence,
> there
> > shouldn't any need for creating tables on segments in utility mode. So,
> > having the behavior of using gp_segment_configuration if info is
> available
> > via it would be good to cover for upgrade and hack won't be needed.
> >
>
> Thanks for the explanation.
>
> I made some experiments again, it's gpcheckcat who connects to segments in
> utility mode. By adding some condition checks and removing some asserts and
> error checks I finally passed the test_gpdb.sh test w/o the pg_upgrade &
> pg_dump changes in this PR.
>
> However my concerns on this approach is that gpcheckcat & pg_upgrade &
> pg_dump are not the only components who access in utility mode, for other
> components whether the hacks are suitable must be carefully checked.
> Considering that we are working on __known__ & __existing__ integration
> issues between gpexpand and other components, I think it is better not to
> introduce new potential integration breakers.
>
> But it's worthy to come back later to this approach.
>
>
Please can you elaborate what problem gpcheckcat is facing when connecting
in utility mode ? Why gpcheckcat is running fine currently in CI after ICW
without this patch ?
Frankly, I lost track of what problem this PR is trying to solve. It stated
was for pg_upgrade, need for which has been void, then would be great to
clearly articulate the new purpose and how its solving the same.

**Comment by pivotal-ning-yu** [comment]
On Fri, Jan 4, 2019 at 3:36 PM Ashwin Agrawal <notifications@github.com>
wrote:

> On Tue, Dec 25, 2018 at 12:58 AM Ning Yu <notifications@github.com> wrote:
>
> > On Mon, Dec 24, 2018 at 8:53 AM Ashwin Agrawal <notifications@github.com
> >
> > wrote:
> >
> > > On Sun, Dec 23, 2018 at 4:46 PM Ning Yu <notifications@github.com>
> > wrote:
> > >
> > > > On Fri, Dec 21, 2018 at 6:38 AM Heikki Linnakangas <
> > > > notifications@github.com>
> > > > wrote:
> > > >
> > > > > If you check that there is no online expansion in progress, and
> > refuse
> > > > the
> > > > > upgrade if there is, do you need the code in pg_dump to modify
> > > > numsegments?
> > > > > There shouldn't be any tables with a different numsegments value
> when
> > > we
> > > > > create the dump, because we already checked for that earlier.
> > > > >
> > > >
> > > > The problem is in binary upgrade mode, in this mode we connect to
> each
> > > > segment in utility mode and restore the schemas, CREATE TABLE sets
> the
> > > > table's numsegments to cluster size, but in cluster size we does not
> > the
> > > > correct cluster size. So far we report cluster size as -1 in utility
> > > mode,
> > > > so all tables' numsegments will be set to 1, however the data are
> > reused
> > > > and were distributed as real cluster size, so there is a mismatch.
> > > >
> > > > One attempt to fix the issue is to let gpdb reports correct cluster
> > size
> > > > even in utility mode. Cluster size is counted from the catalog table
> > > > gp_segment_configuration, however it's only valid on master but empty
> > on
> > > > segments, so there is no way to get the correct count on utility mode
> > > > segments.
> > > >
> > >
> > > In binary upgrade mode, schema is restored only on master first in
> > utility
> > > mode. Then the master's data directory is copied to segments. Hence,
> > there
> > > shouldn't any need for creating tables on segments in utility mode. So,
> > > having the behavior of using gp_segment_configuration if info is
> > available
> > > via it would be good to cover for upgrade and hack won't be needed.
> > >
> >
> > Thanks for the explanation.
> >
> > I made some experiments again, it's gpcheckcat who connects to segments
> in
> > utility mode. By adding some condition checks and removing some asserts
> and
> > error checks I finally passed the test_gpdb.sh test w/o the pg_upgrade &
> > pg_dump changes in this PR.
> >
> > However my concerns on this approach is that gpcheckcat & pg_upgrade &
> > pg_dump are not the only components who access in utility mode, for other
> > components whether the hacks are suitable must be carefully checked.
> > Considering that we are working on __known__ & __existing__ integration
> > issues between gpexpand and other components, I think it is better not to
> > introduce new potential integration breakers.
> >
> > But it's worthy to come back later to this approach.
> >
> >
> Please can you elaborate what problem gpcheckcat is facing when connecting
> in utility mode ? Why gpcheckcat is running fine currently in CI after ICW
> without this patch ?
>

First I want to point out that above descriptions and what I would write
below is about "let gpdb report the actual cluster size in utility mode, so
we do not need to hack pg_dump", which is actually a different approach
from current PR.

gpcheckcat works fine currently when connecting to segments in utility
mode, because we return cluster size as -1 directly in utility mode.  In my
early attempts I hacked the code to count the cluster size from
gp_segment_configuration in utility mode (like what we do on normal
master), it doesn't work on utility mode segments because that catalog
table is empty on segments which do not satisfy some assertions.  Later I
hacked more to return actual cluster size only on normal & utility mode
master, and still return -1 directly on segments (no matter normal mode or
utility mode), then we do make binary upgrade works correctly without the
hacks in pg_dump.

However I still prefer the approach in current PR at the moment.  Unless we
have the other approach fully investigated and verified I have no
confidence on it.


Frankly, I lost track of what problem this PR is trying to solve. It stated
> was for pg_upgrade, need for which has been void, then would be great to
> clearly articulate the new purpose and how its solving the same.
>

The purposes of current PR are not changed all the time:
1. prevent pg_upgrade if gpexpand is in progress;
2. make sure tables' numsegments property is correctly restored in binary
upgrade;

For the first purpose the discussions are focused on __how__ to make the
detection and detect what.  We could detect for gpexpand status files and
schema tables, we could also detect for partial tables.  As the real
concerns are on partial tables current conclusion is to detect for partial
tables directly in pg_upgrade.

For the second purpose, if we accept hacks on pg_dump then we only need to
clarify __when__ to generate the DDL to hack the numsegments property.
There might be other solutions but maybe we could first make it work now
then later make it better.

**Comment by zhangjackey** [comment]
I rollback the PR to the previous commit:
1. Check the `numsegments` of all relations which is in the gp_distribution_policy.
2. Remove the gp_toolkit.__gp_number_of_segments, Now use gp_segment_configuration.

Please help to review, Thanks.

@hlinnaka @pivotal-ning-yu @ashwinstar @danielgustafsson

File: contrib/pg_upgrade/check_gp.c [hunk 6]
```diff
@@ -93,8 +79,10 @@ check_online_expansion(void)
 		res = executeQueryOrDie(conn,
 								"SELECT true AS expansion "
 								"FROM gp_distribution_policy d "
-								"JOIN gp_toolkit.__gp_number_of_segments n "
-								"ON d.numsegments <> n.numsegments "
+								"JOIN (SELECT count(*) segcount "
+								"      FROM gp_segment_configuration "
+								"      WHERE content >= 0 and role = 'm') s "
```
Comment (ashwinstar) [inline contrib/pg_upgrade/check_gp.c]: Why `role = 'm'` and not obvious instead `role = 'p'`?

Comment (zhangjackey) [reply contrib/pg_upgrade/check_gp.c]: Thanks a lot, this is an error, I will correct it.

**Comment by zhangjackey** [comment]
@hlinnaka  @pivotal-ning-yu @ashwinstar @danielgustafsson

This PR is ready, are there any concern? Thanks.

**Comment by hlinnaka** [comment]
Per my earlier comments, the pg_dump changes are no longer necessary.

**Pull request closed without merging**